src/examples/int65.ts

```diff
@@ -0,0 +1,160 @@
+// TODO: review and potentially make this snarkyjs' Int64
+
+import { Field, UInt64, prop, CircuitValue, Circuit, isReady } from 'snarkyjs';
+
+export { Int65 };
+
+class Int65 extends CircuitValue {
+  // * in the range [-2^64+1, 2^64-1], unlike a normal int64
+  // * under- and overflowing is disallowed, similar to UInt64, unlike a normal int64
+
+  @prop magnitude: Field; // absolute value, restricted like UInt64
+  @prop sign: Field; // +/- 1
```

Comment:
Shouldn't we pack the sign bit into the same underlying `Field` corresponding to the magnitude to be more efficient?

Could we instead make this a class with a custom `toFields` rather than a circuit value?

Reply:
I thought about this for a bit now..

`toFields` returns the in-circuit representation, so the main objective is to minimize the number of constraints that result from this representation. Therefore, I think the only candidate for an efficient 1-field representation is the one where the Int65 _is_ the field: `toFields = int65 => [int65.magnitude.mul(int65.sign)]`. Anything else involving bit packing would just lead to very inefficient circuit operations, IIUC.

So, is `magnitude * sign` ("1-field") a more efficient representation than `(magnitude, sign)` ("2-field")?
Several common operations like add, mul, etc, operate on 1-field so in 2-field they result in one additional multiplication constraint. However, the `check` operation (constraining to 64 bits + a sign) which is called at the introduction of _every_ witness, and also at the end of add, mul, etc, operates on 2-field. So here, the 1-field representation needs to add an additional `magnitude * sign = int65` multiplication constraint, which will typically cancel out most of the gains achieved by 1-field elsewhere.
There are some notable operations for which 2-field is definitely better:
* div and mod (which do integer division with rounding on the magnitude)
* converting the Int64 to a `Currency.Amount.Signed` (for the zkapp balance), which has the exact same `(magnitude, sign)` representation we use here. This is currently the main thing an Int64 is used for, and was my original motivation to use 2 fields.

Overall, I think the existing implementation is the optimal one, but happy to be corrected.

Reply:
Sounds good! I love this -- please include this text in a doccomment for the class